Sources/Plasma/Apps/plClient/Mac-Cocoa/.clang-format

```diff
@@ -0,0 +1,48 @@
+---
+BasedOnStyle: Google
+---
+Language: Cpp
+# Modifications to the style for Plasma go here
+IndentWidth: 4
+ColumnLimit: 0
+BraceWrapping:
+  AfterClass:      true
+  AfterControlStatement: MultiLine
+  AfterEnum:       false
+  AfterFunction:   true
+  AfterNamespace:  false
+  AfterObjCDeclaration: false
+  AfterStruct:     true
+  AfterUnion:      false
+  BeforeCatch:     false
+  BeforeElse:      false
+  IndentBraces:    false
+BreakBeforeBraces: Custom
+AlignConsecutiveDeclarations: AcrossComments
+AlignConsecutiveAssignments:
+  Enabled: False
+ReflowComments: false
+LambdaBodyIndentation: OuterScope
+AllowShortBlocksOnASingleLine: true
+AllowShortFunctionsOnASingleLine: true
+AllowShortIfStatementsOnASingleLine: true
+PackConstructorInitializers: CurrentLine
+---
+Language: ObjC
+IndentWidth: 4
+BraceWrapping:
+  AfterClass:      true
+  AfterControlStatement: MultiLine
+  AfterEnum:       false
+  AfterFunction:   true
+  AfterNamespace:  false
+  AfterObjCDeclaration: true
+  AfterStruct:     true
+  AfterUnion:      false
+  BeforeCatch:     false
+  BeforeElse:      false
+  IndentBraces:    false
+BreakBeforeBraces: Custom
+DerivePointerAlignment: False
+PointerAlignment: Right
```

Comment:
Keeping this Clang format file inside of the Mac client directory for now - it could be expanded to other parts of Plasma in the future.

The one thing I couldn't quite get over for Obj-C style is it's very typical for Obj-C to always put the pointer annotation next to the variable name, not the type. I've added a custom rule for Obj-C  to do that.

Unfortunately that could make things a little weird when C++ is mixed in an Obj-C file. Technically that wouldn't match the C++ standards for this project. Unfortunately - I don't think Clang Format is smart enough to differentiate C++ in a Obj-C file.

Reply:
Using clang-format on the new ObjC code is definitely a good idea.

I'm not a fan of formatting Objective-C(++) differently than plain C++ though. The pointer formatting style we use indeed doesn't match the usual ObjC style, but otherwise it works just fine. I would say that consistency within the Plasma codebase is more important than matching what the language/OS does (our pointer style also doesn't match many other C/C++ projects or the Windows headers).

The clang-format config would also be shorter and simpler if we don't use any language-specific styles.

Reply:
I'll consider over the weekend. A few thoughts just off the cuff:
- Plasma is very roughly based on the Google Code Style guide - which suggests right side pointers when in Obj-C. https://google.github.io/styleguide/objcguide.html
- I don't know that I've ever seen Obj-C code with the asterix on the left side in any code style.
- Obj-C is a C language and inherits C conventions. Obj-C++ is an extension, but I should note only some of the Obj-C code in Plasma is Obj-C++.
- I don't know how much crossover there will be between Obj-C and C++ development in Plasma. I'd like to keep the Obj-C comfortable for any Obj-C developers working on the Mac client.

That said - I'm still considering so let me think about it a little bit and poke around to see what other devs are doing.

Reply:
Plasma code style says that pointers bind to the type, so `plMyType* var`.

Reply:
> Plasma code style says that pointers bind to the type, so `plMyType* var`.

The reason Obj-C doesn't follow this convention is that it doesn't have pointer and value types. `NSString*` is redundant because `NSString` is already a pointer type.

Reply:
In since @dpogue mentioned WebKit I checked what they are doing (in since they are another mixed C++/Obj-C code base) and they're walking the same line:
https://github.com/WebKit/WebKit/blob/875c04f6354f5201170ccd9fca83c4af0dff5ff1/.clang-format

Reply:
Not trying to railroad things, just writing things down as I think through them. Obj-C developers have always just kind of accepted that * should go next to the variable name. So I'm thinking about why we do it.

The following is valid C++:
`int* a, b, c;`

It may have an unintentional error (should b and c be pointers?), but it is valid.

`NSString* a, b ,c;`

...is not valid Obj-C. The compiler will not compile this. NSString is a pointer type, so it's not legal to declare b and c as non-pointer types. Which could be one reason Obj-C style always seems to mandate that the pointer should go with the name.

Because value types do not exist in Obj-C (outside of C types), most Obj-C developers don't really strongly know pointer and value semantics and just type the * next to variable names out of habit.

Why does Obj-C do this? If every Obj-C type is a pointer couldn't the compiler just make that assumption and not make you write a * that it will error without anyway? I don't know how it happened. There was a lot of talk about making that change to Obj-C and getting rid of * for Obj-C types in since it's redundant. But instead Apple just replaced the entire language with Swift, which deals with pointer and value semantics completely differently.

Reply:
Yes, Plasma's pointer style doesn't match how C-family languages are actually parsed. I don't really like it, but it *mostly* works out (and I'm not going to convince anyone to change the code style for the entire codebase).

Plasma's answer to the `int* a, b, c;` case is "[don't do that](https://github.com/H-uru/Plasma/wiki/Preferred-Code-Style#type-declarations)" :) This avoids the most common confusing case when attaching the `*` to the type and not the variable.

> NSString is a pointer type, so it's not legal to declare b and c as non-pointer types.

This isn't quite true by the way. The compiler indeed only really allows using ObjC objects through pointers, but `NSString` without a `*` is still its own thing and not a pointer. This is similar to incomplete struct types in plain C/C++: e. g. if you have `class Foo;` with no body, you can't declare a variable `Foo x;`, but you *can* do `Foo* x;`. And in Plasma style, the pointer still goes on the type, even though you can't leave it out here.

Reply:
Sorry, I marked this conversation as resolved and did not mean to. It is not resolved...

I haven't really made a decision here. Was doing more research tonight and thinking it over.

> This isn't quite true by the way. The compiler indeed only really allows using ObjC objects through pointers, but NSString without a * is still its own thing and not a pointer. This is similar to incomplete struct types in plain C/C++: e. g. if you have class Foo; with no body, you can't declare a variable Foo x;, but you can do Foo* x;. And in Plasma style, the pointer still goes on the type, even though you can't leave it out here.

Functionally I'm not sure an NSString without a pointer is anything close to a value type - and is mostly just an oddity carried over from Obj-C being based on C. Notably - Swift deliberately eliminated any sort of way to refer to Obj-C types as value types from the language itself. So it's not even possible to try to type the code for the compiler to error on. The language semantics simply don't have a grammar for referring to a value based Obj-C type.

It's kind of a nonsensical thing in since all Obj-C objects require objc_msg_send, which itself requires a pointer. NSString-not-* is something that can be typed and refers to something - but what it refers to is some undefined type that can't actually exist. It's an Obj-C object that doesn't meet the requirements to be an Obj-C object. An incomplete type still has functional meaning in C++. There is no meaning to a non-pointer type in Obj-C.

I'm not really decided yet and am still looking around at other mixed codebases. It's kind of unfortunate because Obj-C and C++ are essentially both derived from C, so they inherited the same symbols, but functionally they can be very different.

Reply:
Alright - here's where I landed. I will go ahead and switch Obj-C files over to left aligned. The only project I can find that pushes Obj-C into left aligned pointers is Chromium (and related projects like Electron.) So it is done in practice by at least one major project.

That said:
- Most major projects follow the WebKit style of C++ files being left aligned and Obj-C files being right aligned.
- Right aligned is the Apple standard - so any Apple tooling (e.g. Xcode) will attempt to enforce right alignment in various Obj-C tooling like automatic protocol conformance, outlet/action creation, etc. I do recognize Plasma is also sometimes fighting Intellisense on Windows. But this will be ongoing friction.
- Obj-C is a different language than C++, and I'm still not sure it should automatically be held to the same code standards. (I.E. line length enforcement seems problematic for Obj-C as discussed elsewhere in this thread because Obj-C and C++ have different standards for how functions are named.) Just because C++ can be embedded in Obj-C should not automatically mean they're treated as the same language.

Longer term - the Obj-C will probably be replaced with Swift anyway, which is why I'm probably not going to wrestle with this change too much. An iOS/Vision Pro client will likely get a new SwiftUI based client. Swift/C++ interop is shipping later this year, and thats really the needed missing piece.

My guess is the Obj-C/AppKit client will always be here in some form, and may be what the Mac ships with for the foreseeable future. (@dpogue certainly may find it useful for PPC Mac work.) But I am working on a Swift/SwiftUI client for other platforms. And Swift happens to sidestep issues like this pointer alignment issue by not being C derived at all.

Reply:
> * Right aligned is the Apple standard - so any Apple tooling (e.g. Xcode) will attempt to enforce right alignment in various Obj-C tooling

Hrm, this is unfortunate... If that only happens occasionally and clang-format fixes it, I would still say that we should use "\* on the type" and stay consistent with the C++ code. But if this gets in the way of normal editing in Xcode, it's a different story...

Reply:
It's an annoyance that just means going back and editing Xcode's suggestions. But I can't complain too much because Intellisense on the Windows/Visual Studio side will also suggest things that don't fit Plasma code style. Plasma doesn't align with IDE completion standards just in general - which is ok. Xcode and Visual Studio could disagree on C++ in theory.

Little more worried about what it could mean for any sample Apple implementations brought in (or in theory Metal-cpp) but as long as thats dealt with as contributed code, should be fine.

Reply:
> Little more worried about what it could mean for any sample Apple implementations brought in (or in theory Metal-cpp) but as long as thats dealt with as contributed code, should be fine.

My preference would still be that we find a way to consume metal-cpp as a library rather than vendoring it in-repo, but Apple seems to not want to support that (very obvious) use case